src/west/cmd/project.py

```diff
@@ -480,56 +480,67 @@ def _manifest_path(args):
 
 
 def _fetch(project):
-    # Clone 'project' if it does not exist, and 'git fetch' it otherwise. Also
-    # update the 'manifest-rev' branch to point to the revision specified in
-    # the manifest.
+    # Fetches upstream changes for 'project' and updates the 'manifest-rev'
+    # branch to point to the revision specified in the manifest. If the
+    # project's repository does not already exist, it is created first.
     #
-    # Returns True if the project already existed, and False if it was cloned.
+    # Returns True if the project's repository already existed.
 
-    if _cloned(project):
-        existed = True
-        _inf(project, 'Fetching changes for (name-and-path)')
-        _git(project, 'remote update')
-    else:
-        existed = False
+    exists = _cloned(project)
+
+    if not exists:
+        _inf(project, 'Creating repository for (name-and-path)')
+        _git_base(project, 'init (abspath)')
+        _git(project, 'remote add origin (url)')
 
-        # --depth implies --single-branch, so we must pass --branch when
-        # creating a shallow clone (since the remote might not have HEAD
-        # pointing to the 'revision' branch).
+    if project.clone_depth:
+        _inf(project,
+             'Fetching changes for (name-and-path) with --depth (clone-depth)')
+
+        # If 'clone-depth' is specified, fetch just the specified revision
+        # (probably a branch). That will download the minimum amount of data,
+        # which is probably what's wanted whenever 'clone-depth is used. The
+        # default 'git fetch' behavior is to do a shallow clone of all branches
+        # on the remote.
         #
-        # Currently, we also pass --branch for non-shallow clones, which checks
-        # the 'revision' branch out in the newly created repo, but this might
-        # change.
-
-        msg = 'Cloning (name-and-path)'
-        cmd = 'clone'
-        if _is_sha(project.revision):
-            # If the project revision is a SHA and the remote repo's
-            # HEAD doesn't point to anything valid, the clone
-            # operation might fail to check anything out. In that
-            # case, make sure to check out the given revision.
-            checkout = True
-        else:
-            cmd += ' --branch (revision)'
-            checkout = False
-        if project.clone_depth:
-            msg += ' with --depth (clone-depth)'
-            cmd += ' --depth (clone-depth)'
-        cmd += ' (url) (path)'
+        # Note: Many servers won't allow fetching arbitrary commits by SHA.
+        # Combining --depth with an SHA will break for those.
 
-        _inf(project, msg)
-        _git_base(project, cmd)
+        # Qualify branch names with refs/heads/, just to be safe. Just the
+        # branch name is likely to work as well though.
```

Comment:
Seems worth testing this with a repository that combines clone-depth with a SHA revision as well

Reply:
Would break for most servers I think, because they don't allow fetching an arbitrary SHA (though it works if you already have the commit and it's reachable from some branch/tag you've fetched).

Reply:
Agreed, thanks for the references and extra digging